driver-core/src/main/com/mongodb/internal/connection/EventHelper.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.connection.ClusterDescription;
+import com.mongodb.connection.ServerDescription;
+
+import java.util.Objects;
+
+final class EventHelper {
+
+    static boolean equivalentEvent(final ClusterDescription current, final ClusterDescription previous) {
+        if (!exceptionsEquals(current.getSrvResolutionException(), previous.getSrvResolutionException())) {
+            return false;
+        }
+        if (current.getServerDescriptions().size() != previous.getServerDescriptions().size()) {
+            return false;
+        }
+        for (ServerDescription curNew: current.getServerDescriptions()) {
+            ServerDescription matchingPrev = null;
+            for (ServerDescription curPrev: previous.getServerDescriptions()) {
+                if (curNew.getAddress().equals(curPrev.getAddress())) {
+                    matchingPrev = curPrev;
+                    break;
+                }
+            }
+            if (!equivalentEvent(curNew, matchingPrev)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    static boolean equivalentEvent(final ServerDescription current, final ServerDescription previous) {
```

Comment:
Yeah, the naming choice was deliberate, but I see why it's confusing. I didn't want to call it `equivalentDescriptions`, since that would beg the question of why we're not just using `ServerDescription#equals`.  It's really more like 'wouldDescriptionsGenerateEquivalentEvents'.  Would that be better?  Or perhaps just a comment explaining it? Both?

Reply:
I think both would be good - it looks like a typo now and future Ross would probably wonder why we're not using equals!

Reply:
Done